arangod/Scheduler/ThreadPoolScheduler.cpp

```diff
@@ -0,0 +1,123 @@
+
+void ThreadPoolScheduler::toVelocyPack(velocypack::Builder& b) const {
+  b.add("scheduler-threads", VPackValue(0));  // numWorkers
+  b.add("blocked", VPackValue(0));            // obsolete
+  b.add("queued", VPackValue(0));             // scheduler queue length
+  b.add("in-progress",
+        VPackValue(0));                 // number of working (non-idle) threads
+  b.add("direct-exec", VPackValue(0));  // obsolete
+}
+Scheduler::QueueStatistics ThreadPoolScheduler::queueStatistics() const {
+  return QueueStatistics();
+}
+void ThreadPoolScheduler::trackCreateHandlerTask() noexcept {}
+void ThreadPoolScheduler::trackBeginOngoingLowPriorityTask() noexcept {
+  _metrics->_ongoingLowPriorityGauge += 1;
+}
+void ThreadPoolScheduler::trackEndOngoingLowPriorityTask() noexcept {
+  _metrics->_ongoingLowPriorityGauge -= 1;
+}
+void ThreadPoolScheduler::trackQueueTimeViolation() {}
+void ThreadPoolScheduler::trackQueueItemSize(std::int64_t int64) noexcept {}
+
+uint64_t ThreadPoolScheduler::getLastLowPriorityDequeueTime() const noexcept {
+  return _lastLowPriorityDequeueTime.load(std::memory_order::relaxed);
+}
+
+void ThreadPoolScheduler::setLastLowPriorityDequeueTime(
+    uint64_t time) noexcept {
+  _lastLowPriorityDequeueTime.store(time, std::memory_order::relaxed);
+}
+
+std::pair<uint64_t, uint64_t>
+ThreadPoolScheduler::getNumberLowPrioOngoingAndQueued() const {
+  auto dequeued =
+      _threadPools[int(RequestPriority::LOW)]->statistics.dequeued.load();
+  auto queued =
+      _threadPools[int(RequestPriority::LOW)]->statistics.queued.load();
+  auto remaining = dequeued > queued ? 0 : queued - dequeued;
+  return std::make_pair(
+      _metrics->_ongoingLowPriorityGauge.load(std::memory_order_relaxed),
+      remaining);
+}
+
+double ThreadPoolScheduler::approximateQueueFillGrade() const { return 0; }
+
+double ThreadPoolScheduler::unavailabilityQueueFillGrade() const { return 0; }
+
+bool ThreadPoolScheduler::queueItem(RequestLane lane,
+                                    std::unique_ptr<WorkItemBase> item,
+                                    bool bounded) {
+  item->enqueueTime = std::chrono::steady_clock::now();
+  auto prio = PriorityRequestLane(lane);
+  _threadPools[int(prio)]->push(std::move(item));
+  return true;
+}
+
+ThreadPoolScheduler::ThreadPoolScheduler(
+    ArangodServer& server, uint64_t maxThreads,
+    std::shared_ptr<SchedulerMetrics> metrics)
+    : Scheduler(server), _metrics(std::move(metrics)) {
+  ThreadPoolMetrics poolMetrics = {
+      .jobsDone = &_metrics->_metricsJobsDoneTotal,
+      .jobsQueued = &_metrics->_metricsJobsSubmittedTotal,
+      .jobsDequeued = &_metrics->_metricsJobsDequeuedTotal,
+  };
+
+  _threadPools.reserve(4);
+
+  poolMetrics.queueLength = _metrics->_metricsQueueLengths[0];
+  poolMetrics.dequeueTimes = _metrics->_metricsDequeueTimes[0];
+  _threadPools.emplace_back(std::make_unique<WorkStealingThreadPool>(
+      "SchedMaintenance", std::max(std::ceil(maxThreads * 0.1), 2.),
```

Comment:
nit: the `std::max` here returns (and in the following occasions) returns a double value, but the constructor argument for `WorkStealingThreadPool` is a `size_t` `threadCount`. While this works, an explicit cast to size_t may be more readable here.